// zip-lib/lib/util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unlink = unlink;
exports.mkdir = mkdir;
exports.realpath = realpath;
exports.stat = stat;
exports.lstat = lstat;
exports.chmod = chmod;
exports.readdir = readdir;
exports.access = access;
exports.rmdir = rmdir;
exports.symlink = symlink;
exports.readlink = readlink;
const fs = require("fs");
const util = require("util");
function unlink(path) {
    return util.promisify(fs.unlink)(path);
}
function mkdir(path, mode) {
    return util.promisify(fs.mkdir)(path, mode);
}
function realpath(path) {
    return util.promisify(fs.realpath)(path);
}
function stat(path) {
    return util.promisify(fs.stat)(path);
}
function lstat(path) {
    return util.promisify(fs.lstat)(path);
}
function chmod(path, mode) {
    return util.promisify(fs.chmod)(path, mode);
}
function readdir(path) {
    return util.promisify(fs.readdir)(path);
}
function access(path, mode) {
    return util.promisify(fs.access)(path, mode);
}
function rmdir(path) {
    return util.promisify(fs.rmdir)(path);
}
function symlink(target, path, type) {
    return util.promisify(fs.symlink)(target, path, type);
}
function readlink(path) {
    return util.promisify(fs.readlink)(path);
}
